Default: Generalise, optimise and simplify grass spread function

paramat opened the pull request:
Credit to tenplus1 for the suggestion to generalise for mod use.
Mods can add mod nodes to 'group:spreading_dirt_type' enabling the
function to work with mod nodes.

Add some nodes to this group.

Removing 'dirt_with_grass' etc. from 'neighbors' stops the ABM action
running everywhere and constantly, on the dirt nodes immediately below
the surface nodes. Now the action only runs in the rare case of a dirt
node with neighbouring air, grass decorations or snow.

Remove check for air above to allow grass to spread under light-
transmitting nodes such as fences, walls, plants. This causes spread
under slabs, stairs and glass, when near air, but seems worth it.

Remove unnecessary check for nil node.
/////////////////////////////////////////////////////

![screenshot_20160929_210539](https://cloud.githubusercontent.com/assets/3686677/18970319/b773025e-8688-11e6-9279-458f0f3f0d67.png)

Partly derived from #1286 and including the optimisation i discovered during that discussion.
Thanks to tenplus1.

The light check takes care of 'ignore' above.
If 'unknown' above the groups checks simply fail.

paramat commented:
Tested.

sofar commented:
Yeah, this is what I was thinking about as well.

The only thing I am contemplating is to define a new group `seeds = 1` that could be put into the nodedef of any plant, and have the plant nodedef have a function or a member that declares what node to seed dirt into.

e.g.

```
def = {
  name = "farming:grass_1",
  groups = { seeds = 1, grass = 1, choppy = 2 },
  seeds = "default:dirt_with_grass",
  ...
```

This way, we don't have to hard code that `default:snow` seeds `dirt` to `dirt_with_snow`, and mods can use it trivially. We can then also remove the `group:grass` and `group:dry_grass` from the ABM and use `group:seeds` instead, simplifying it further.

tenplus1 commented:
@sofar - you are over-complicating what should be a simple spreading function.

paramat commented:
sofar possibly a good idea, can we consider that in another PR?

Ferk commented:
An alternative would be to have a `default.dirt_seeders` table much like the `default.dirt_types` you added.

And then use `default.dirt_seeders` as metatable for the table to be used as `neighbors` of the abm.

```
default.dirt_seeders= { "group:grass", "group:dry_grass", "default:snow" }
local grass_spread_neighbors = {"air"}
setmetatable(grass_spread_neighbors, default.dirt_seeders)

minetest.register_abm({
 	label = "Grass spread",
 	label = "Grass spread",
 	nodenames = {"default:dirt"},
 	nodenames = {"default:dirt"},
 	neighbors = {
 	neighbors = grass_spread_neighbors,
        ...
```

paramat commented:
t4im suggested using a new node group instead of a table of spreading nodes, worth considering.

paramat commented:
Updatd to use a group.

t4im commented on mods/default/functions.lua:
```diff
@@ -366,8 +364,7 @@ minetest.register_abm({
 		-- Look for likely neighbors.
-		local p2 = minetest.find_node_near(pos, 1, {"default:dirt_with_grass",
-				"default:dirt_with_dry_grass", "default:dirt_with_snow"})
+		local p2 = minetest.find_node_near(pos, 1, "group:spreads_on_dirt")
```
Groups stand in for items, so this groups "anything, that is spreading on dirt", like flowers, grass, but not other dirt_with_grass imo.
And with flowers this might actually delete nodes. :-)

How about `grassy_dirt` or  alternatively just `dirt` with a node definition property check of `spreads_on_dirt` after?

paramat replied:
Yeah the group name probably needs improvement.

paramat replied:
Hm 'grassy_dirt_spread' .. 'spreading_grassy_dirt'
I'm trying to be more generalised as the surface may not be grassy.
'spreading_dirt_node' ..

paramat replied:
'spread_dirty_node'

paramat replied:
spreading_dirt_type ..

rubenwardy commented:
Is there any way of benchmarking ABMs?

tenplus1 commented:
I did a for loop (1000 iterations) for current, paramat's and my own changes and the above function was faster.

paramat commented:
Interesting, your PR has a light check, a get node for a water check (unnecessary), and 'find nodes in area' (often slower than 'find node near').
Mine has a light check, 'find node near' and a get node for decorations above.
For many actions 'find node near' will search the entire volume because it doesn't find anything.
So i think our PRs will be similar in speed, and it looks like they are.
However mine keeps an additional feature needed for seeding isolated areas of dirt.

paramat commented:
Updated to use 'spreading_dirt_type'.
Also use that group in the grass death ABM.

tenplus1 commented:
@paramat - like I mentioned in my own pull, you NEED a water check otherwise grass will grow under the edges of flowing water next to air...  It may speed things up by removing it entirely but results in glitched grass spreading in places it should not go.

Ferk commented on mods/default/functions.lua:
```diff
@@ -366,8 +364,7 @@ minetest.register_abm({
+		local p2 = minetest.find_node_near(pos, 1, "group:spreading_dirt_type")
 		if p2 then
 			-- But the node needs to be under air in this case.
 			local n2 = minetest.get_node(above)
```
I know it's not really the purpose of this PR, but `minetest.get_node(above)` is done twice in the case p2 exists but it's not air. That second call after this if block wouldn't really be needed if you get n2 already before you even check for p2.
That would be more optimal.

paramat replied:
Will attempt.

Ferk replied:
actually, you could check if n2 is not "air" and in that case you wouldn't even need to do the `minetest.find_node_near`.

paramat replied:
Well, it's highly likely the dirt will be under air as it has 'neighbors = air, grasses, snow'. But your suggestion will simplify the code.

paramat replied:
Both ideas seem good.

paramat commented:
> grass will grow under the edges of flowing water next to air

Ok, but that's a rare occurence (doesn't happen in high-altitude rivers), and usually only happens when water is placed and flows over grass.

paramat commented:
What is this check for, only for ignore above?
```
		-- Anything on top?
		local n2 = minetest.get_node(above)
		if not n2 then
			return
		end
```
Not sure that's needed as ignore would return zero light earlier causing a return?

paramat commented:
Updated and much simplified.
I removed the check for air above for spreading, because we would want dirt under, for example, a fence or other non-light-blocking node, to spread.
I removed the check mentioned above, it should have been 'node or nil' and ignore above would have caused a return in the light check.

paramat commented:
I might re-add that check for air above.

paramat commented:
Yes the air check is needed, grass spreads under glass, slabs and stairs.

paramat commented:
Updated.

paramat commented:
Just seen another optimisation, if air is above we don't need to then check for seeding nodes above.

paramat commented:
Done.
This just gets better and better.

tenplus1 commented:
So wait, dirt under fences, gates, torches will never turn into grass ?

paramat commented:
Unfortunately correct, how would you distinguish them from nodeboxes which have a high internal light value?

Ferk commented:
There could be an "allow_dirt_spread" group. I'm not sure if it's worth it, though.

@sofar suggestion of using "group:seeds_dirt" would simplify the group checks though. From 2 groups and 1 node.name to just 1 group check.

tenplus1 commented:
Same way I do on my own pull #1286 by checking above for ligth level < 13 and only checking node above for water/ignore.

Ferk commented:
@tenplus1 then dirt will grow below slabs and stairs.

tenplus1 commented:
That's ok by me, what if slab is on it's side and *some* light gets through?

paramat commented:
Good points.

paramat commented:
I'll do this but won't insist on it.

paramat commented:
Updated to spread under light-transmitting nodes, see screenshot in first post.

sofar commented:
If it spreads under fences and stuff like in the screenshot (I'm assuming it was all dirt nodes before) is fine, I like it. Very consistent.

:+1:

paramat commented:
Yes the rows were all dirt, then seeded in the middle and spread outwards, stopped by the light-blocking node only.

rubenwardy commented on mods/default/functions.lua:
```diff
@@ -348,9 +348,7 @@ minetest.register_abm({
-		"default:dirt_with_grass",
-		"default:dirt_with_dry_grass",
-		"default:dirt_with_snow",
+		"air",
```
why `air` instead of `group:spreading_dirt_type`?

paramat replied:
This is the big optimisation mentioned in tenplus1's PR:
"Removing 'dirt_with_grass' etc. from 'neighbors' stops the ABM action
running everywhere and constantly, on the dirt nodes immediately below
the surface nodes. Now the action only runs in the rare case of a dirt
node with neighbouring air, grass decorations or snow."
If 'group:spreading_dirt_type' was included in neighbours, normal unaltered biomes would trigger a large number of ABM actions on the dirt nodes just below the surface, each action being a light check which would 'return' due to darkness.
So now the slow Lua ABM action only runs when spreading is needed.
'air' is needed in neighbours to trigger for exposed dirt.

rubenwardy replied:
I see. Sounds good

paramat commented:
c0de5646d230b209af32c60c3e5d1256855b2767
